beacon_node/beacon_chain/src/data_column_verification.rs

```diff
@@ -449,29 +437,30 @@ where
 /// sent bad data. The re-verification cost should not be significant. If a peer sends invalid data it
 /// will be quickly banned.
 pub fn verify_kzg_for_data_column_list_with_scoring<'a, E: EthSpec, I>(
-    data_column_iter: I,
+    mut data_column_iter: I,
     kzg: &'a Kzg,
-) -> Result<(), Vec<(ColumnIndex, KzgError)>>
+) -> Result<(), Option<(ColumnIndex, KzgError)>>
 where
     I: Iterator<Item = &'a Arc<DataColumnSidecar<E>>> + Clone,
 {
     if verify_kzg_for_data_column_list(data_column_iter.clone(), kzg).is_ok() {
         return Ok(());
     };
 
-    // Find all columns that are invalid and identify by index. If we hit this condition there
-    // should be at least one invalid column
-    let errors = data_column_iter
-        .filter_map(|data_column| {
-            if let Err(e) = verify_kzg_for_data_column(data_column.clone(), kzg) {
-                Some((data_column.index, e))
-            } else {
-                None
-            }
-        })
-        .collect::<Vec<_>>();
+    // Find the first column that is invalid and identify by index. If we hit this condition, there
+    // should be at least one invalid column.
+    // Only finding the first invalid column because of the chance of multiple peers sending invalid
+    // columns for the same batch should be very rare, and re-verifying all columns individually is
+    // costly with likely little benefit.
+    let error = data_column_iter.find_map(|data_column| {
```

Comment:
If we chunk it by columns indices, we can complete this function in one single pass in the failure case